core/src/main/java/org/apache/iceberg/rest/HTTPHeaders.java

```diff
@@ -0,0 +1,168 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableListMultimap;
+import org.apache.iceberg.relocated.com.google.common.collect.ListMultimap;
+import org.apache.iceberg.relocated.com.google.common.collect.Multimap;
+import org.immutables.value.Value;
+
+/**
+ * Represents a group of HTTP headers.
+ *
+ * <p>Header name comparison in this class is always case-insensitive, in accordance with RFC 2616.
+ *
+ * <p>This class exposes methods to convert to and from different representations such as maps and
+ * multimap, for easier access and manipulation – especially when dealing with multiple headers with
+ * the same name.
+ */
+@Value.Style(depluralize = true)
+@Value.Immutable
+@SuppressWarnings({"ImmutablesStyle", "SafeLoggingPropagation"})
+public interface HTTPHeaders {
+
+  HTTPHeaders EMPTY = of();
+
+  /** Returns all the header entries in this group. */
+  List<HTTPHeader> entries();
+
+  /**
+   * Returns a map representation of the headers where each header name is mapped to a list of its
+   * values. Header names are case-insensitive.
+   */
+  @Value.Lazy
+  default Map<String, List<String>> asMap() {
+    return entries().stream()
+        .collect(Collectors.groupingBy(h -> h.name().toLowerCase(Locale.ROOT)))
+        .values()
+        .stream()
+        .collect(
+            Collectors.toMap(
+                headers -> headers.get(0).name(),
+                headers -> headers.stream().map(HTTPHeader::value).collect(Collectors.toList())));
+  }
+
+  /**
+   * Returns a simple map representation of the headers where each header name is mapped to its
+   * first value. If a header has multiple values, only the first value is used. Header names are
+   * case-insensitive.
+   */
+  @Value.Lazy
+  default Map<String, String> asSimpleMap() {
+    return entries().stream()
+        .collect(Collectors.groupingBy(h -> h.name().toLowerCase(Locale.ROOT)))
+        .values()
+        .stream()
+        .collect(
+            Collectors.toMap(headers -> headers.get(0).name(), headers -> headers.get(0).value()));
+  }
+
+  /**
+   * Returns a {@link ListMultimap} representation of the headers. Header names are
+   * case-insensitive.
+   */
+  @Value.Lazy
+  default ListMultimap<String, String> asMultiMap() {
```

Comment:
do we actually have a use case for this? I see the value in having `asMap()` / `asSimpleMap()` but I feel like we don't need this method atm?

Reply:
This one will be very useful later to do things like this in `HTTPClient`:

```java
HTTPRequest req = ...
HttpUriRequestBase request = new HttpUriRequestBase(req.method().name(), req.requestUri());
req.headers().asMultiMap().forEach(request::addHeader);
```